beacon/light/committee_chain.go

```diff
@@ -0,0 +1,587 @@
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	log.Info("Inserted new committee update", "period", period, "next committee root", update.NextSyncCommitteeRoot)
+	return nil
+}
+
+// NextSyncPeriod returns the next period where an update can be added and also
+// whether the chain is initialized at all.
+func (s *CommitteeChain) NextSyncPeriod() (uint64, bool) {
+	s.chainmu.RLock()
+	defer s.chainmu.RUnlock()
+
+	if s.committees.periods.IsEmpty() {
+		return 0, false
+	}
+	if !s.updates.periods.IsEmpty() {
+		return s.updates.periods.AfterLast, true
+	}
+	return s.committees.periods.AfterLast - 1, true
+}
+
+// rollback removes all committees and fixed roots from the given period and updates
+// starting from the previous period.
+func (s *CommitteeChain) rollback(batch ethdb.Batch, period uint64) {
+	s.deleteCommitteesFrom(batch, period)
+	s.fixedRoots.deleteFrom(batch, period)
+	if period > 0 {
+		period--
+	}
+	s.updates.deleteFrom(batch, period)
+}
+
+// getCommitteeRoot returns the committee root at the given period, either fixed,
+// proven by a previous update or both. It returns an empty hash if the committee
+// root is unknown.
+func (s *CommitteeChain) getCommitteeRoot(period uint64) common.Hash {
+	if root, ok := s.fixedRoots.get(period); ok || period == 0 {
+		return root
+	}
+	if update, ok := s.updates.get(period - 1); ok {
+		return update.NextSyncCommitteeRoot
+	}
+	return common.Hash{}
+}
+
+// getSyncCommittee returns the deserialized sync committee at the given period.
+func (s *CommitteeChain) getSyncCommittee(period uint64) syncCommittee {
+	if c, ok := s.syncCommitteeCache.Get(period); ok {
+		return c
+	}
+	if sc, ok := s.committees.get(period); ok {
+		c, err := s.sigVerifier.deserializeSyncCommittee(sc)
+		if err != nil {
+			log.Error("Sync committee deserialization error", "error", err)
+			return nil
+		}
+		s.syncCommitteeCache.Add(period, c)
+		return c
+	}
+	log.Error("Missing serialized sync committee", "period", period)
+	return nil
+}
+
+// VerifySignedHeader returns true if the given signed header has a valid signature
+// according to the local committee chain. The caller should ensure that the
+// committees advertised by the same source where the signed header came from are
+// synced before verifying the signature.
+// The age of the header is also returned (the time elapsed since the beginning
+// of the given slot, according to the local system clock). If enforceTime is
+// true then negative age (future) headers are rejected.
+func (s *CommitteeChain) VerifySignedHeader(head types.SignedHeader) (bool, time.Duration) {
+	s.chainmu.RLock()
+	defer s.chainmu.RUnlock()
+
+	return s.verifySignedHeader(head)
+}
+
+func (s *CommitteeChain) verifySignedHeader(head types.SignedHeader) (bool, time.Duration) {
+	var age time.Duration
+	now := s.unixNano()
+	if head.Header.Slot < (uint64(now-math.MinInt64)/uint64(time.Second)-s.config.GenesisTime)/12 {
```

Comment:
why to minus `MinInt64` ...

Reply:
Btw, I think calculating the expected time/age of a slot can be something common. Maybe to have a common package `beacon/time` or something else. Just a nitpick.

Reply:
This check ensures that the calculated age will not exceed the `time.Duration` (`int64`)  range in the negative direction. It backwards calculates the `Slot` value at which the age would be `MinInt64`. Hence the minus `MinInt64`.

Reply:
This is currently the only place where the absolute time of a slot is used so I'm not sure about adding a separate package for it just yet. If there are more general purpose beacon chain related functions then maybe it's justified but I currently cannot think of anything else to put there.